torch/csrc/cuda/nccl.cpp

```diff
@@ -0,0 +1,175 @@
+#include "nccl.h"
+#include "torch/csrc/THP.h"
+#include "torch/csrc/Types.h"
+#include "torch/csrc/DynamicTypes.h"
+#include "torch/csrc/cuda/THCP.h"
+
+#include <nccl.h>
+#include <sstream>
+#include <unordered_map>
+
+static inline void CHECK(ncclResult_t status)
+{
+  if (status != ncclSuccess) {
+    std::stringstream err;
+    err << "NCCL Error " << status << ": " << ncclGetErrorString(status);
+    throw std::runtime_error(err.str());
+  }
+}
+
+// TODO: make this thread_local + add mutexes
+std::unordered_map<std::string, ncclComm_t*> _communicators;
+
+PyObject * THCPModule_nccl_destroy(PyObject *args) {
+  for(auto it: _communicators) {
+    ncclCommDestroy(*(it.second));
+  }
+  Py_RETURN_NONE;
+}
+
+static ncclComm_t* _get_communicator(std::vector<at::Tensor>& inputs, int ndevices) {
+  std::stringstream hash;
+  std::vector<int> devs;
+  for (int i = 0; i < ndevices; i++) {
+    int dev = inputs[i].get_device();
+    hash <<  dev << ",";
+    devs.push_back(dev);
+  }
+  if (_communicators.find(hash.str()) == _communicators.end()) {
+    ncclComm_t *comms = (ncclComm_t*) malloc(sizeof(ncclComm_t) * ndevices);
+    if (comms == NULL) throw std::runtime_error("failed to allocate ncclComm_t*");
+    CHECK(ncclCommInitAll(comms, ndevices, &devs[0]));
+    _communicators[hash.str()] = comms;
+  }
+  return _communicators[hash.str()];
+}
+
+static void _check_inputs(std::vector<at::Tensor> &inputs, std::vector<at::Tensor> &outputs, int size_multiplier) {
```

Comment:
what is this `size_multiplier` for?

Reply:
it's not for reduce, but for other operations. as i implement more nccl calls, it'll be used.